include/alpaka/mem/buf/BufCudaRt.hpp

```diff
@@ -118,12 +118,13 @@ namespace alpaka
                 {
                     ALPAKA_DEBUG_MINIMAL_LOG_SCOPE;
 
-                    // Set the current device. \TODO: Is setting the current device before cudaFree required?
+                    // Set the current device.
```

Comment:
Are you sure that `cudaSetDevice` is really necessary for `cudaFree`?

Reply:
`cudaMalloc` and `cudaFree` handle device dependent memory, so `cudaSetDevice` should be used in both cases. There is no documentation, that `cudaFree` is context-agnostic. It is true, that `cudaFree` is smart enough to find the correct device for the given memory address, if UVA is present (64-bit, TCC; however not tested the non-UVA case).

I have experienced in this [program](https://github.com/tdd11235813/read-nvml-clocks-pci), that `cudaFree` acts on the currently set device. The example uses a simple `cudaFree(0)` after `cudaSetDevice`, where last one does only lazy context creation/initialization but is a low-overhead call. `cudaFree` uses a given device context and if the device context is wrong, application clock settings do not apply (e.g. set by `nvidia-smi -ac`). If no `cudaFree(0)` is used, clock settings also do not apply at all. You then have to first run a warmup-kernel. It is not a real error, but it shows that cudaFree uses and writes context information.